File: python/hostpay/resources.py
"""Resource groups mapped to the HostPay money surface.

Each method mirrors a real endpoint; see ../openapi.json for the full API.
"""
from __future__ import annotations

from typing import Any, Optional

from .models import EscrowResponse, TransactionResponse, UserRead, WalletRead

# Mobile-money providers (wire values expected by the API).
PROVIDER_ORANGE = "m17"
PROVIDER_AFRICELL = "m18"


class _Resource:
    def __init__(self, transport: Any) -> None:
        self._t = transport


class Users(_Resource):
    def create(
        self,
        app_user_id: str,
        name: str,
        phone_number: str,
        email: Optional[str] = None,
        username: Optional[str] = None,
    ) -> UserRead:
        return self._t.request("POST", "/api/v1/users/create/", json={
            "app_user_id": app_user_id,
            "name": name,
            "phone_number": phone_number,
            "email": email,
            "username": username,
        })

    def get(self, user_id: str) -> UserRead:
        return self._t.request("GET", f"/api/v1/users/{user_id}/")

    def list(self, is_active: Optional[bool] = None) -> list[UserRead]:
        params = {} if is_active is None else {"is_active": is_active}
        return self._t.request("GET", "/api/v1/users/", params=params)

    def update(
        self,
        user_id: str,
        app_user_id: str,
        name: str,
        phone_number: str,
        email: Optional[str] = None,
        username: Optional[str] = None,
    ) -> UserRead:
        """Full update — the API expects the complete user body; app_user_id
        must match the existing value (it is immutable)."""
        return self._t.request("PUT", f"/api/v1/users/{user_id}/", json={
            "app_user_id": app_user_id,
            "name": name,
            "phone_number": phone_number,
            "email": email,
            "username": username,
        })


    def patch(
        self,
        user_id: str,
        name: Optional[str] = None,
        email: Optional[str] = None,
        username: Optional[str] = None,
        phone_number: Optional[str] = None,
        is_active: Optional[bool] = None,
    ) -> UserRead:
        """Partial update — only the fields you pass are changed
        (app_user_id is immutable)."""
        body = {
            k: v
            for k, v in {
                "name": name,
                "email": email,
                "username": username,
                "phone_number": phone_number,
                "is_active": is_active,
            }.items()
            if v is not None
        }
        return self._t.request("PATCH", f"/api/v1/users/{user_id}/", json=body)

    def delete(self, user_id: str) -> Any:
        return self._t.request("DELETE", f"/api/v1/users/{user_id}/")

    def disable(self, user_id: str) -> Any:
        return self._t.request("POST", f"/api/v1/users/{user_id}/disable")

    def enable(self, user_id: str) -> Any:
        return self._t.request("POST", f"/api/v1/users/{user_id}/enable")


class Wallets(_Resource):
    def create(self, user_id: str) -> WalletRead:
        return self._t.request("POST", f"/api/v1/wallets/create/{user_id}/")

    def get(self, user_id: str) -> WalletRead:
        return self._t.request("GET", f"/api/v1/wallets/{user_id}/")

    def balance(self, wallet_id: str) -> Any:
        return self._t.request("GET", f"/api/v1/wallets/{wallet_id}/balance")

    def list(self, is_active: Optional[bool] = None) -> list[WalletRead]:
        params = {} if is_active is None else {"is_active": is_active}
        return self._t.request("GET", "/api/v1/wallets/", params=params)

    def disable(self, wallet_id: str) -> Any:
        return self._t.request("POST", f"/api/v1/wallets/{wallet_id}/disable")

    def enable(self, wallet_id: str) -> Any:
        return self._t.request("POST", f"/api/v1/wallets/{wallet_id}/enable")


class Transactions(_Resource):
    def get(self, transaction_id: str) -> TransactionResponse:
        return self._t.request("GET", f"/api/v1/transactions/{transaction_id}")

    def list(
        self,
        status: Optional[str] = None,
        transaction_type: Optional[str] = None,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        search: Optional[str] = None,
        limit: int = 100,
        offset: int = 0,
    ) -> list[TransactionResponse]:
        params = {
            "status": status,
            "transaction_type": transaction_type,
            "start_date": start_date,
            "end_date": end_date,
            "search": search,
            "limit": limit,
            "offset": offset,
        }
        return self._t.request(
            "GET",
            "/api/v1/transactions/",
            params={k: v for k, v in params.items() if v is not None},
        )

    def for_wallet(self, wallet_id: str) -> list[TransactionResponse]:
        """All transactions for a wallet, incoming and outgoing."""
        return self._t.request("GET", f"/api/v1/transactions/wallet/{wallet_id}")

    def sync(self, reference_id: str) -> Any:
        """Trigger an immediate reconciliation sync for one of your
        transactions by its provider reference id — useful right after a
        payment completes."""
        return self._t.request("POST", f"/api/v1/transactions/sync/{reference_id}")


class Deposits(_Resource):
    def mobile_money(
        self, wallet_id: str, amount: int, idempotency_key: Optional[str] = None
    ) -> Any:
        return self._t.request(
            "POST",
            "/api/v1/transactions/wallet/mobile-money-deposit",
            json={"wallet_id": wallet_id, "amount": amount},
            idempotency_key=idempotency_key,
        )

    def card(
        self,
        wallet_id: str,
        amount: float,
        payment_method_id: Optional[str] = None,
        idempotency_key: Optional[str] = None,
    ) -> Any:
        return self._t.request(
            "POST",
            "/api/v1/transactions/wallet/card-deposit/create",
            json={
                "wallet_id": wallet_id,
                "amount": amount,
                "payment_method_id": payment_method_id,
            },
            idempotency_key=idempotency_key,
        )


class Transfers(_Resource):
    def create(
        self,
        sender_wallet_id: str,
        recipient_identifier: str,
        amount: float,
        description: Optional[str] = None,
        idempotency_key: Optional[str] = None,
    ) -> TransactionResponse:
        return self._t.request(
            "POST",
            "/api/v1/transactions/wallet/transfer/",
            json={
                "sender_wallet_id": sender_wallet_id,
                "recipient_identifier": recipient_identifier,
                "amount": amount,
                "description": description,
            },
            idempotency_key=idempotency_key,
        )


class Payouts(_Resource):
    def mobile_money(
        self,
        wallet_id: str,
        amount: float,
        phone_number: str,
        provider: str = PROVIDER_ORANGE,
        currency: str = "SLE",
        idempotency_key: Optional[str] = None,
    ) -> TransactionResponse:
        return self._t.request(
            "POST",
            "/api/v1/transactions/wallet/mobile-money-cashout/",
            json={
                "wallet_id": wallet_id,
                "amount": amount,
                "phone_number": phone_number,
                "provider": provider,
                "currency": currency,
            },
            idempotency_key=idempotency_key,
        )

    def bank(
        self,
        wallet_id: str,
        amount: float,
        currency: str = "usd",
        description: Optional[str] = None,
        idempotency_key: Optional[str] = None,
    ) -> TransactionResponse:
        return self._t.request(
            "POST",
            "/api/v1/transactions/wallet/payout/",
            json={
                "wallet_id": wallet_id,
                "amount": amount,
                "currency": currency,
                "description": description,
            },
            idempotency_key=idempotency_key,
        )


class Escrow(_Resource):
    def hold(
        self,
        wallet_id: str,
        amount: float,
        description: Optional[str] = None,
        idempotency_key: Optional[str] = None,
    ) -> EscrowResponse:
        return self._t.request(
            "POST",
            "/api/v1/escrow/hold",
            json={
                "wallet_id": wallet_id,
                "amount": amount,
                "description": description,
            },
            idempotency_key=idempotency_key,
        )

    def release(
        self,
        transaction_id: str,
        recipient_wallet_id: str,
        amount: Optional[float] = None,
        idempotency_key: Optional[str] = None,
    ) -> EscrowResponse:
        return self._t.request(
            "POST",
            f"/api/v1/escrow/{transaction_id}/release",
            json={"recipient_wallet_id": recipient_wallet_id, "amount": amount},
            idempotency_key=idempotency_key,
        )

    def refund(
        self,
        transaction_id: str,
        amount: Optional[float] = None,
        idempotency_key: Optional[str] = None,
    ) -> EscrowResponse:
        return self._t.request(
            "POST",
            f"/api/v1/escrow/{transaction_id}/refund",
            json={"amount": amount},
            idempotency_key=idempotency_key,
        )


class Fees(_Resource):
    def summary(self) -> Any:
        return self._t.request("GET", "/api/v1/fees/summary")

    def configuration(self) -> Any:
        return self._t.request("GET", "/api/v1/fees/configuration")

    def estimate_deposit(
        self, amount: float, payment_method: str, is_international: bool = False
    ) -> Any:
        """payment_method: 'mobile_money' | 'card' | 'bank' | 'wallet'."""
        return self._t.request("POST", "/api/v1/fees/estimate/deposit", json={
            "amount": amount,
            "payment_method": payment_method,
            "is_international": is_international,
        })

    def estimate_withdrawal(self, amount: float, payment_method: str) -> Any:
        return self._t.request("POST", "/api/v1/fees/estimate/withdrawal", json={
            "amount": amount,
            "payment_method": payment_method,
        })

    def estimate_transfer(self, amount: float) -> Any:
        return self._t.request("POST", "/api/v1/fees/estimate/transfer", json={
            "amount": amount,
        })

    def estimate_card_metadata(self, payment_method_id: str, amount: float) -> Any:
        """Card-aware deposit estimate using the actual card's country/brand."""
        return self._t.request(
            "POST", "/api/v1/fees/estimate/deposit/card-metadata", json={
                "payment_method_id": payment_method_id,
                "amount": amount,
            }
        )


class Testing(_Resource):
    def simulate_monime_webhook(
        self, transaction_id: str, status: str = "successful"
    ) -> Any:
        """Complete or fail a pending Test Mode mobile-money deposit.
        Test keys only — the API rejects this in Live Mode. status:
        'successful' | 'failed'."""
        return self._t.request(
            "POST", "/api/v1/testing/simulate-monime-webhook", json={
                "transaction_id": transaction_id,
                "status": status,
            }
        )


class WebhookSubscriptions(_Resource):
    def create(
        self,
        target_url: str,
        events: list[str],
        description: Optional[str] = None,
        ip_allowlist: Optional[list[str]] = None,
        payload_version: Optional[str] = None,
    ) -> Any:
        """Create a subscription. The response includes the signing secret
        ONCE — store it; it cannot be retrieved again (only rotated)."""
        body: dict = {"target_url": target_url, "events": list(events)}
        if description is not None:
            body["description"] = description
        if ip_allowlist is not None:
            body["ip_allowlist"] = list(ip_allowlist)
        if payload_version is not None:
            body["payload_version"] = payload_version
        return self._t.request("POST", "/api/v1/webhooks/subscriptions", json=body)

    def list(self) -> Any:
        return self._t.request("GET", "/api/v1/webhooks/subscriptions")

    def update(
        self,
        subscription_id: str,
        target_url: Optional[str] = None,
        events: Optional[list[str]] = None,
        active: Optional[bool] = None,
        ip_allowlist: Optional[list[str]] = None,
        payload_version: Optional[str] = None,
    ) -> Any:
        body = {
            k: v
            for k, v in {
                "target_url": target_url,
                "events": events,
                "active": active,
                "ip_allowlist": ip_allowlist,
                "payload_version": payload_version,
            }.items()
            if v is not None
        }
        return self._t.request(
            "PATCH", f"/api/v1/webhooks/subscriptions/{subscription_id}", json=body
        )

    def delete(self, subscription_id: str) -> Any:
        return self._t.request(
            "DELETE", f"/api/v1/webhooks/subscriptions/{subscription_id}"
        )

    def rotate_secret(self, subscription_id: str) -> Any:
        """Returns the new signing secret once."""
        return self._t.request(
            "POST", f"/api/v1/webhooks/subscriptions/{subscription_id}/rotate-secret"
        )


class Connect(_Resource):
    """Stripe Connect onboarding for user payout accounts."""

    def complete_onboarding(
        self,
        wallet_id: str,
        individual: dict,
        business_profile: dict,
        client_ip: str,
        card_token: Optional[str] = None,
    ) -> Any:
        """Submit identity + business details for the wallet owner's Connect
        account.

        client_ip must be the END CUSTOMER's IP address — Stripe records it
        as evidence of Terms-of-Service acceptance. Passing your server's IP
        is a compliance violation, so the SDK refuses to guess it.
        """
        if not client_ip:
            raise ValueError(
                "client_ip is required: Stripe records it for TOS acceptance "
                "and it must be the end customer's IP, not your server's"
            )
        return self._t.request(
            "POST",
            "/api/v1/transactions/wallet/complete-onboarding/",
            json={
                "wallet_id": wallet_id,
                "individual": individual,
                "business_profile": business_profile,
                "card_token": card_token,
            },
            headers={"X-Forwarded-For": client_ip},
        )

    def upload_verification_document(
        self,
        wallet_id: str,
        document: Any,
        document_side: str,
        filename: str = "document.jpg",
        mime_type: str = "image/jpeg",
    ) -> Any:
        """Upload an identity document (JPEG/PNG/PDF, max 10 MB).

        document: bytes or a file-like object. document_side: 'front' | 'back'.
        """
        return self._t.request(
            "POST",
            f"/api/v1/transactions/wallet/{wallet_id}/connect/verification-document",
            data={"document_side": document_side},
            files={"file": (filename, document, mime_type)},
        )

    def status(self, wallet_id: str) -> Any:
        """Sync and return the Connect account's verification status."""
        return self._t.request(
            "GET", f"/api/v1/transactions/wallet/{wallet_id}/connect/status"
        )

    def delete(self, wallet_id: str) -> Any:
        return self._t.request(
            "POST",
            "/api/v1/transactions/wallet/connect/delete",
            params={"wallet_id": wallet_id},
        )

File: typescript/src/resources.ts
/** Resource groups mapped to the HostPay money surface. See ../openapi.json. */
import type { Escrow as EscrowModel, HostPayObject, Transaction, User, Wallet } from "./models.js";

export interface Transport {
  request(
    method: string,
    path: string,
    init?: {
      body?: unknown;
      idempotencyKey?: string;
      query?: Record<string, unknown>;
      headers?: Record<string, string>;
      formData?: FormData;
    },
  ): Promise<any>;
}

/** Mobile-money providers (wire values the API expects). */
export const Provider = {
  orange: "m17",
  africell: "m18",
} as const;
export type ProviderCode = (typeof Provider)[keyof typeof Provider];

abstract class Resource {
  constructor(protected t: Transport) {}
}

export class Users extends Resource {
  create(params: {
    appUserId: string;
    name: string;
    phoneNumber: string;
    email?: string;
    username?: string;
  }): Promise<User> {
    return this.t.request("POST", "/api/v1/users/create/", {
      body: {
        app_user_id: params.appUserId,
        name: params.name,
        phone_number: params.phoneNumber,
        email: params.email,
        username: params.username,
      },
    });
  }

  get(userId: string): Promise<User> {
    return this.t.request("GET", `/api/v1/users/${userId}/`);
  }

  list(params: { isActive?: boolean } = {}): Promise<User[]> {
    return this.t.request("GET", "/api/v1/users/", {
      query: { is_active: params.isActive },
    });
  }

  /** Full update — the API expects the complete user body; appUserId must
   * match the existing value (it is immutable). */
  update(
    userId: string,
    params: {
      appUserId: string;
      name: string;
      phoneNumber: string;
      email?: string;
      username?: string;
    },
  ): Promise<User> {
    return this.t.request("PUT", `/api/v1/users/${userId}/`, {
      body: {
        app_user_id: params.appUserId,
        name: params.name,
        phone_number: params.phoneNumber,
        email: params.email,
        username: params.username,
      },
    });
  }

  /** Partial update — only the fields you pass are changed (appUserId is immutable). */
  patch(
    userId: string,
    params: {
      name?: string;
      email?: string;
      username?: string;
      phoneNumber?: string;
      isActive?: boolean;
    },
  ): Promise<User> {
    const body: Record<string, unknown> = {};
    if (params.name !== undefined) body.name = params.name;
    if (params.email !== undefined) body.email = params.email;
    if (params.username !== undefined) body.username = params.username;
    if (params.phoneNumber !== undefined) body.phone_number = params.phoneNumber;
    if (params.isActive !== undefined) body.is_active = params.isActive;
    return this.t.request("PATCH", `/api/v1/users/${userId}/`, { body });
  }

  delete(userId: string): Promise<any> {
    return this.t.request("DELETE", `/api/v1/users/${userId}/`);
  }

  disable(userId: string): Promise<any> {
    return this.t.request("POST", `/api/v1/users/${userId}/disable`);
  }

  enable(userId: string): Promise<any> {
    return this.t.request("POST", `/api/v1/users/${userId}/enable`);
  }
}

export class Wallets extends Resource {
  create(userId: string): Promise<Wallet> {
    return this.t.request("POST", `/api/v1/wallets/create/${userId}/`);
  }

  get(userId: string): Promise<Wallet> {
    return this.t.request("GET", `/api/v1/wallets/${userId}/`);
  }

  balance(walletId: string): Promise<HostPayObject> {
    return this.t.request("GET", `/api/v1/wallets/${walletId}/balance`);
  }

  list(params: { isActive?: boolean } = {}): Promise<Wallet[]> {
    return this.t.request("GET", "/api/v1/wallets/", {
      query: { is_active: params.isActive },
    });
  }

  disable(walletId: string): Promise<any> {
    return this.t.request("POST", `/api/v1/wallets/${walletId}/disable`);
  }

  enable(walletId: string): Promise<any> {
    return this.t.request("POST", `/api/v1/wallets/${walletId}/enable`);
  }
}

export class Transactions extends Resource {
  get(transactionId: string): Promise<Transaction> {
    return this.t.request("GET", `/api/v1/transactions/${transactionId}`);
  }

  list(
    params: {
      status?: string;
      transactionType?: string;
      startDate?: string;
      endDate?: string;
      search?: string;
      limit?: number;
      offset?: number;
    } = {},
  ): Promise<Transaction[]> {
    return this.t.request("GET", "/api/v1/transactions/", {
      query: {
        status: params.status,
        transaction_type: params.transactionType,
        start_date: params.startDate,
        end_date: params.endDate,
        search: params.search,
        limit: params.limit,
        offset: params.offset,
      },
    });
  }

  /**
   * Trigger an immediate reconciliation sync for one of your transactions by
   * its provider reference id — useful right after a payment completes.
   */
  sync(referenceId: string): Promise<any> {
    return this.t.request("POST", `/api/v1/transactions/sync/${referenceId}`);
  }

  /** All transactions for a wallet, incoming and outgoing. */
  forWallet(walletId: string): Promise<Transaction[]> {
    return this.t.request("GET", `/api/v1/transactions/wallet/${walletId}`);
  }
}

export class Deposits extends Resource {
  mobileMoney(params: {
    walletId: string;
    amount: number;
    idempotencyKey?: string;
  }): Promise<HostPayObject> {
    return this.t.request(
      "POST",
      "/api/v1/transactions/wallet/mobile-money-deposit",
      {
        body: { wallet_id: params.walletId, amount: params.amount },
        idempotencyKey: params.idempotencyKey,
      },
    );
  }

  card(params: {
    walletId: string;
    amount: number;
    paymentMethodId?: string;
    idempotencyKey?: string;
  }): Promise<HostPayObject> {
    return this.t.request(
      "POST",
      "/api/v1/transactions/wallet/card-deposit/create",
      {
        body: {
          wallet_id: params.walletId,
          amount: params.amount,
          payment_method_id: params.paymentMethodId,
        },
        idempotencyKey: params.idempotencyKey,
      },
    );
  }
}

export class Transfers extends Resource {
  create(params: {
    senderWalletId: string;
    recipientIdentifier: string;
    amount: number;
    description?: string;
    idempotencyKey?: string;
  }): Promise<Transaction> {
    return this.t.request("POST", "/api/v1/transactions/wallet/transfer/", {
      body: {
        sender_wallet_id: params.senderWalletId,
        recipient_identifier: params.recipientIdentifier,
        amount: params.amount,
        description: params.description,
      },
      idempotencyKey: params.idempotencyKey,
    });
  }
}

export class Payouts extends Resource {
  mobileMoney(params: {
    walletId: string;
    amount: number;
    phoneNumber: string;
    provider?: ProviderCode;
    currency?: string;
    idempotencyKey?: string;
  }): Promise<Transaction> {
    return this.t.request(
      "POST",
      "/api/v1/transactions/wallet/mobile-money-cashout/",
      {
        body: {
          wallet_id: params.walletId,
          amount: params.amount,
          phone_number: params.phoneNumber,
          provider: params.provider ?? Provider.orange,
          currency: params.currency ?? "SLE",
        },
        idempotencyKey: params.idempotencyKey,
      },
    );
  }

  bank(params: {
    walletId: string;
    amount: number;
    currency?: string;
    description?: string;
    idempotencyKey?: string;
  }): Promise<Transaction> {
    return this.t.request("POST", "/api/v1/transactions/wallet/payout/", {
      body: {
        wallet_id: params.walletId,
        amount: params.amount,
        currency: params.currency ?? "usd",
        description: params.description,
      },
      idempotencyKey: params.idempotencyKey,
    });
  }
}

export class Escrow extends Resource {
  hold(params: {
    walletId: string;
    amount: number;
    description?: string;
    idempotencyKey?: string;
  }): Promise<EscrowModel> {
    return this.t.request("POST", "/api/v1/escrow/hold", {
      body: {
        wallet_id: params.walletId,
        amount: params.amount,
        description: params.description,
      },
      idempotencyKey: params.idempotencyKey,
    });
  }

  release(
    transactionId: string,
    params: { recipientWalletId: string; amount?: number; idempotencyKey?: string },
  ): Promise<EscrowModel> {
    return this.t.request("POST", `/api/v1/escrow/${transactionId}/release`, {
      body: {
        recipient_wallet_id: params.recipientWalletId,
        amount: params.amount,
      },
      idempotencyKey: params.idempotencyKey,
    });
  }

  refund(
    transactionId: string,
    params: { amount?: number; idempotencyKey?: string } = {},
  ): Promise<EscrowModel> {
    return this.t.request("POST", `/api/v1/escrow/${transactionId}/refund`, {
      body: { amount: params.amount },
      idempotencyKey: params.idempotencyKey,
    });
  }
}

export class Fees extends Resource {
  summary(): Promise<any> {
    return this.t.request("GET", "/api/v1/fees/summary");
  }

  configuration(): Promise<any> {
    return this.t.request("GET", "/api/v1/fees/configuration");
  }

  /** paymentMethod: "mobile_money" | "card" | "bank" | "wallet". */
  estimateDeposit(params: {
    amount: number;
    paymentMethod: string;
    isInternational?: boolean;
  }): Promise<any> {
    return this.t.request("POST", "/api/v1/fees/estimate/deposit", {
      body: {
        amount: params.amount,
        payment_method: params.paymentMethod,
        is_international: params.isInternational ?? false,
      },
    });
  }

  estimateWithdrawal(params: { amount: number; paymentMethod: string }): Promise<any> {
    return this.t.request("POST", "/api/v1/fees/estimate/withdrawal", {
      body: { amount: params.amount, payment_method: params.paymentMethod },
    });
  }

  estimateTransfer(params: { amount: number }): Promise<any> {
    return this.t.request("POST", "/api/v1/fees/estimate/transfer", {
      body: { amount: params.amount },
    });
  }

  /** Card-aware deposit estimate using the actual card's country/brand. */
  estimateCardMetadata(params: { paymentMethodId: string; amount: number }): Promise<any> {
    return this.t.request("POST", "/api/v1/fees/estimate/deposit/card-metadata", {
      body: { payment_method_id: params.paymentMethodId, amount: params.amount },
    });
  }
}

export class Testing extends Resource {
  /**
   * Complete or fail a pending Test Mode mobile-money deposit. Test keys
   * only — the API rejects this in Live Mode.
   */
  simulateMonimeWebhook(params: {
    transactionId: string;
    status?: "successful" | "failed";
  }): Promise<any> {
    return this.t.request("POST", "/api/v1/testing/simulate-monime-webhook", {
      body: {
        transaction_id: params.transactionId,
        status: params.status ?? "successful",
      },
    });
  }
}

export class WebhookSubscriptions extends Resource {
  /**
   * Create a subscription. The response includes the signing secret ONCE —
   * store it; it cannot be retrieved again (only rotated).
   */
  create(params: {
    targetUrl: string;
    events: string[];
    description?: string;
    ipAllowlist?: string[];
    payloadVersion?: string;
  }): Promise<any> {
    const body: Record<string, unknown> = {
      target_url: params.targetUrl,
      events: params.events,
    };
    if (params.description !== undefined) body.description = params.description;
    if (params.ipAllowlist !== undefined) body.ip_allowlist = params.ipAllowlist;
    if (params.payloadVersion !== undefined) body.payload_version = params.payloadVersion;
    return this.t.request("POST", "/api/v1/webhooks/subscriptions", { body });
  }

  list(): Promise<any> {
    return this.t.request("GET", "/api/v1/webhooks/subscriptions");
  }

  update(
    subscriptionId: string,
    params: {
      targetUrl?: string;
      events?: string[];
      active?: boolean;
      ipAllowlist?: string[];
      payloadVersion?: string;
    },
  ): Promise<any> {
    const body: Record<string, unknown> = {};
    if (params.targetUrl !== undefined) body.target_url = params.targetUrl;
    if (params.events !== undefined) body.events = params.events;
    if (params.active !== undefined) body.active = params.active;
    if (params.ipAllowlist !== undefined) body.ip_allowlist = params.ipAllowlist;
    if (params.payloadVersion !== undefined) body.payload_version = params.payloadVersion;
    return this.t.request("PATCH", `/api/v1/webhooks/subscriptions/${subscriptionId}`, { body });
  }

  delete(subscriptionId: string): Promise<void> {
    return this.t.request("DELETE", `/api/v1/webhooks/subscriptions/${subscriptionId}`);
  }

  /** Returns the new signing secret once. */
  rotateSecret(subscriptionId: string): Promise<any> {
    return this.t.request("POST", `/api/v1/webhooks/subscriptions/${subscriptionId}/rotate-secret`);
  }
}

export class Connect extends Resource {
  /**
   * Submit identity + business details for the wallet owner's Stripe Connect
   * account. `clientIp` must be the END CUSTOMER's IP address — Stripe records
   * it as evidence of Terms-of-Service acceptance, so the SDK refuses to guess it.
   */
  completeOnboarding(params: {
    walletId: string;
    individual: Record<string, unknown>;
    businessProfile: Record<string, unknown>;
    clientIp: string;
    cardToken?: string;
  }): Promise<any> {
    if (!params.clientIp) {
      throw new Error(
        "clientIp is required: Stripe records it for TOS acceptance and it must be the end customer's IP, not your server's",
      );
    }
    return this.t.request("POST", "/api/v1/transactions/wallet/complete-onboarding/", {
      body: {
        wallet_id: params.walletId,
        individual: params.individual,
        business_profile: params.businessProfile,
        card_token: params.cardToken,
      },
      headers: { "X-Forwarded-For": params.clientIp },
    });
  }

  /** Upload an identity document (JPEG/PNG/PDF, max 10 MB). */
  uploadVerificationDocument(params: {
    walletId: string;
    document: Blob | Uint8Array;
    documentSide: "front" | "back";
    filename?: string;
    mimeType?: string;
  }): Promise<any> {
    const blob =
      params.document instanceof Blob
        ? params.document
        : new Blob([params.document as Uint8Array<ArrayBuffer>], { type: params.mimeType ?? "image/jpeg" });
    const formData = new FormData();
    formData.set("document_side", params.documentSide);
    formData.set("file", blob, params.filename ?? "document.jpg");
    return this.t.request(
      "POST",
      `/api/v1/transactions/wallet/${params.walletId}/connect/verification-document`,
      { formData },
    );
  }

  /** Sync and return the Connect account's verification status. */
  status(walletId: string): Promise<any> {
    return this.t.request("GET", `/api/v1/transactions/wallet/${walletId}/connect/status`);
  }

  delete(walletId: string): Promise<any> {
    return this.t.request("POST", "/api/v1/transactions/wallet/connect/delete", {
      query: { wallet_id: walletId },
    });
  }
}
